src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -724,53 +805,140 @@ public PrefixEnumerable FindKeysWithPrefix(string prefix)
-                _prefix = prefix;
+                ValidationState = entry.ValidationState;
+                Visible = true;
             }
 
-            public PrefixEnumerator GetEnumerator()
+            public void Reset()
             {
-                return _dictionary == null ? new PrefixEnumerator() : new PrefixEnumerator(_dictionary, _prefix);
+                Visible = false;
+                RawValue = null;
+                AttemptedValue = null;
+                ValidationState = ModelValidationState.Unvalidated;
+                Errors.Clear();
             }
 
-            IEnumerator<KeyValuePair<string, ModelStateEntry>>
-                IEnumerable<KeyValuePair<string, ModelStateEntry>>.GetEnumerator()
+            public ModelStateNode GetNode(StringSegment subKey, bool createIfNotExists)
             {
-                return GetEnumerator();
+                if (subKey.Length == 0)
+                {
+                    return this;
+                }
+
+                var index = BinarySearch(subKey);
+                ModelStateNode modelStateNode = null;
+                if (index >= 0)
+                {
+                    modelStateNode = ChildNodes[index];
+                }
+                else if (createIfNotExists)
+                {
+                    if (ChildNodes == null)
+                    {
+                        ChildNodes = new List<ModelStateNode>(1);
+                    }
+
+                    modelStateNode = new ModelStateNode(this, subKey);
+                    ChildNodes.Insert(~index, modelStateNode);
+                }
+
+                return modelStateNode;
             }
 
-            IEnumerator IEnumerable.GetEnumerator()
+            private int BinarySearch(StringSegment searchKey)
             {
-                return GetEnumerator();
+                if (ChildNodes == null)
+                {
+                    return -1;
+                }
+
+                var low = 0;
+                int high = ChildNodes.Count - 1;
+                while (low <= high)
+                {
+                    var mid = low + ((high - low) / 2);
+                    var midKey = ChildNodes[mid].SubKey;
+
+                    // Nodes in the tree are stored with the . delimiter (for e.g. ".foo" would be the subkey).
+                    // However, we'd like the ability to query a property by it's name. Skip the '.' if the query
+                    // subkey does not start with one.
+                    var midKeyOffset = searchKey.Buffer[searchKey.Offset] != '.' && midKey.Buffer[midKey.Offset] == '.'
+                        ? midKey.Offset + 1
+                        : midKey.Offset;
+
+                    var result = string.Compare(
+                        midKey.Buffer,
+                        midKeyOffset,
+                        searchKey.Buffer,
+                        searchKey.Offset,
+                        Math.Max(midKey.Length, searchKey.Length),
```

Comment:
Never mind. I misread this code and though it was using `Math.Min()`. That would mean "extra" characters were ignored. As it's actually written, all is fine because `Compare()` won't go past the end of either `string`.